### PR TITLE
Improve specular light in phong shading

### PR DESCRIPTION
**Pull request by almarklein**
Addresses the main point in  #2087

* Make the depth step for calculating the view direction a bit smaller to avoid glitches.
* Fixes that the view direction was reversed. I don't think this caused any problems, but it did cause confusion while making changes. (I verified the direction by adding `light_vec = eye_vec`.)
* Changes the reflection model to the better Blinn-Phong: see https://learnopengl.com/Advanced-Lighting/Advanced-Lighting.
* Tweaked default light paramseters.

Dropping this here:
```
    // Note (AK): Below is a way to flip the surface normal so that it's always in
    // the view direction. However, this only works well when the light and view
    // vector are (about) the same. Otherwise, normals on faces that that are
    // near-parallel to the eye vector can flip, affecting how they're lit,
    // causing bright and dark artifacts at the edges of objects.
    normal = faceforward(-normal, normal, eye_vec);
```

### CODE REVIEW

**Comment by almarklein** [comment]
I hoped to also introduce the `faceforward` call here, which makes that the shader is more forgiving about the orientation of surface normals. However, due to how the eye vector is calculated, this causes artifacts. I started changing the eye vector math, but then thought there might be a reason for it being this way.

**Comment by djhoese** [comment]
CC @asnt

**Comment by asnt** [comment]
This seems to fix the random glittering that was appearing on the tubes in `examples/basics/visual/tube.py` and I don't see a visual regression on the other mesh examples.

File: vispy/visuals/filters/mesh.py
Comment (asnt) [inline vispy/visuals/filters/mesh.py:144]: So, the main change is here, right? The eye vector is created in normalised coordinates and is thus now independent of the size of the scene/frustum/models. Right?

File: vispy/visuals/filters/mesh.py
Comment (asnt) [inline vispy/visuals/filters/mesh.py:138]: I am not sure I understand what the plane is here.

File: vispy/visuals/filters/mesh.py
Comment (asnt) [inline vispy/visuals/filters/mesh.py:197]: Ah ok so this is the plane.

I'm not sure we want to systematically orient normals towards the viewer. This might correct some issues with inconsistent face orientations of the input models (#2069), but in specific cases only. For example, for face culling to work, there is no other possibility than to input a geomertry with a consistent definition.

**Review by asnt — COMMENTED** [review]
This seems to fix one issue with the 'glittering' effect in `examples/basics/visuals/tube.py` following #2087. That was the main point of this PR, right? Or is there something else that this PR solves that I didn't notice?

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [reply vispy/visuals/filters/mesh.py:144]: The main change is actually turning those `1e6` into `1e5`. There is a bit of re-ordering to do the math with less transforms .

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [reply vispy/visuals/filters/mesh.py:197]: This description is also not correct. To be honest I am not sure why the artifacts appear.

File: vispy/visuals/filters/mesh.py
Comment (asnt) [reply vispy/visuals/filters/mesh.py:144]: I see, changing to `1e5` only suffices, but I still have no evidence that this is not specific to this example. I wonder if a perspective correction would not be a more general solution. Examples:

No perspective correction (`1e6` too small relative to the size of the viewing frustum):
```c
    // Fails: 'glittering' artefacts when not close enough to the tubes in `examples/basics/visuals/tube.py`.
    vec4 pos_eye = $scene2doc(pos_scene);

    vec4 pos_front = pos_eye;
    pos_front.z += 1e-6;
    pos_front = $doc2scene(pos_front);
    pos_front /= pos_front.w;

    vec4 pos_back = pos_eye;
    pos_back.z -= 1e-6;
    pos_back = $doc2scene(pos_back);
    pos_back /= pos_back.w;
```

No perspective correction (`1e5` is big enough):
```c
    // Works.
    vec4 pos_eye = $scene2doc(pos_scene);

    vec4 pos_front = pos_eye;
    pos_front.z += 1e-5;
    pos_front = $doc2scene(pos_front);
    pos_front /= pos_front.w;

    vec4 pos_back = pos_eye;
    pos_back.z -= 1e-5;
    pos_back = $doc2scene(pos_back);
    pos_back /= pos_back.w;
```

With perspective correction, variant 1:
```c
    // Works.
    vec4 pos_eye = $scene2doc(pos_scene);

    pos_eye /= pos_eye.w;

    vec4 pos_front = pos_eye;
    pos_front.z += 1e-6;
    pos_front = $doc2scene(pos_front);
    pos_front /= pos_front.w;

    vec4 pos_back = pos_eye;
    pos_back.z -= 1e-6;
    pos_back = $doc2scene(pos_back);
    pos_back /= pos_back.w;
```

With perspective correction, variant 2 (the math is equivalent to variant 1):
```c
    // Works.
    vec4 pos_eye = $scene2doc(pos_scene);

    vec4 pos_front = pos_eye;
    pos_front.z += 1e-6 * pos_front.w;
    pos_front = $doc2scene(pos_front);
    pos_front /= pos_front.w;

    vec4 pos_back = pos_eye;
    pos_back.z -= 1e-6 * pos_back.w;
    pos_back = $doc2scene(pos_back);
    pos_back /= pos_back.w;
```

I'll try to check that in actual example tests.

Comment (asnt) [reply vispy/visuals/filters/mesh.py:144]: To summarise, I think this change makes things more correct and would be good to merge.

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [inline vispy/visuals/filters/mesh.py:212]: Changed from Phong to Blinn-Phong. If I had not done that I'd only have to change `reflect(-light_vec, normal);` (which is what you see in online code examples).

**Comment by almarklein** [comment]
This is ready from my end. I added notes to the top post.

**Comment by djhoese** [comment]
As we work on the new CI failures, I think this should be included in 0.7, right? @asnt if you have time to re-review that'd be great. Let's include this one, the `render` alpha changes, and my CoC PR and then release this thing. ...maybe have some napari folks test it.

File: vispy/visuals/filters/mesh.py
Comment (asnt) [reply vispy/visuals/filters/mesh.py:140]: I think this is really the only change needed to fix the main point. Isn't it? (Besides, clarifying the definition of `$light_dir`.)

**Review by asnt — CHANGES_REQUESTED** [review]
I have the impression that there more changes than needed to fix the main point ("glitches"/"glitter artefact").
Could we reduce the focus to the main point and keep the rest for separate PR's?

Except for one point: The remaining ambiguity here for me is the definition of [`MeshVisual.light_dir`](https://github.com/vispy/vispy/blob/2eea22fbb0d4eea82351aca2dddca1785d3db242/vispy/visuals/mesh.py#L482). Is it the direction
- to the light source?
- or away from the light source? (i.e. the direction of propagation of the rays).

Intuitively, since it is a directional light, I'd say it is the direction of propagation. Given, the docstring I'd say it is the direction of propagation, too. But given how the diffuse component is computed `dot(light_vec, normal)` (and given that `light_vec = light_dir`), I'd say that `light_dir` is used as if it was the direction to the light source.
AFAICT, this is the only parameters exposed publicly that affects the choicr of signs inside the shader. So, if we can clarify this for this PR, then we make the minimal amount of changes to account for it, i.e. change the signs of `light_dir/light_vec`appropriately wherever they appear. Then, in another PR, we clarify the internal implementation by consistently defining the signs of the other direction vectors.

For the rest, good points are made:
- OK for the new default light intensities, but this is per se a different issue.
- If think the Phong model works and is standard enough to keep as the default. Blinn-Phong is a nice addition, but I'd rather have as an alternative (it's known to be more efficient and I've read it renders the highlights better in some ways, but I have no evidence that it is that noticeable. And if it is, it deserves being an alternative instead.)

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [reply vispy/visuals/filters/mesh.py:212]: 1. As I understand it, Blinn-Phong avoids issues with Phong at certain angles, so it its is considered an improvement. I only know because I read this, so correct me if I'm wrong :)   but if this is the case it seems like a better choice then, right?
2. I see your point. I tested it by making the light_vec and eye_vec equal, but the light_vec might not mean what I thought it meant. Let's add these definitions as comments in the shader. I think I'm going to take a look at this tomorrow morning, it's too late for this now :)

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [reply vispy/visuals/filters/mesh.py:142]: +z is away from the camera in NDC. I think the same applied to doc as well, but maybe double-check...

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [reply vispy/visuals/filters/mesh.py:193]: Yeah, I should remove it. The fact that this caused artifacts while I have this working fine elsewhere really annoyed me and. Now that I understand, I want to document/remember it. But this is not the place.

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [reply vispy/visuals/filters/mesh.py:399]: No, the reason is that because it already saturates quite a bit with the diffuse light, the specular light is barely visible.

File: vispy/visuals/filters/mesh.py
Comment (almarklein) [reply vispy/visuals/filters/mesh.py:140]: This fixes the glitches, yes. Some of the other changes further improve/tweak the (specular) lighting, so still apply to this PR, I guess.

**Comment by almarklein** [comment]
Updated and squashed.

I have commented the direction of the vectors. For both light_vec and eye_vec I've defined them to point towards the surface. This is surely the most natural for the light source. For the eye_vec there is more room for discussion, but this felt more natural to me (e.g. thinking of ray-tracing in volume rendering). 

It seems @asnt and I disagree about Phong vs Blinn-Phong for the reflection math. I don't see a reason to chose (or even support) Phong if Blinn-Phong is considered better. But I also don't feel too strongly about it to let this hold up this PR.

**Comment by almarklein** [comment]
Also updated the mesh shading test, since the light now comes from the other (correct) direction. Great feeling that that test broke by this change, actually.

File: vispy/visuals/filters/mesh.py
Comment (djhoese) [inline vispy/visuals/filters/mesh.py:183]: Just curious, does this perform differently in the shader or is this change purely style?

Comment (almarklein) [reply vispy/visuals/filters/mesh.py:183]: You want to avoid branches in shaders, and a ternary operation is more trivial to convert into branch-free code by the compiler.

**Comment by djhoese** [comment]
Ok so here's my summary of what I understand from the few comments I've read here:

1. @asnt implemented Phong, @almarklein implemented Blinn-Phong.
2. According to wikipedia Blinn-Phong produces a better result. This makes me agree with @almarklein that I'm not sure why we should have Phong if Blinn-Phong can also be implemented. And I'm not sure there is a reason to support both unless someone is specifically comparing the two results (not something we need to support).
3. There seems to be some confusion over conventions with things like light direction and other properties, but docstrings were added to help minimize the confusion. None of the implementation is necessarily wrong (it produces valid results), but maynot necessarily match certain conventions depending on who you talk to and what paper you read.
4. The current implementation in this PR is Phong with Blinn-Phong commented out, right @almarklein ?

Where do we go from here? I'm in favor of Blinn-Phong if it produces better results and is already implemented but commented out. That said, I don't know what I'm talking about. I'd like to get this merged and then request that some napari folks double check a few things with VisPy main (meshes, visuals, images, shading, etc) so that we can still release tomorrow (if napari folks have a ton of free time) or as late as early next week.

**Comment by almarklein** [comment]
> The current implementation in this PR is Phong with Blinn-Phong commented out, right @almarklein ?

yes

**Comment by djhoese** [comment]
@almarklein With the way Blinn-Phong is implemented here, does it still have the higher efficiency that wikipedia mentions? Do you have a screenshot of the difference these two implementations will make in VisPy?

If the difference with which one of these is implemented comes down to a few lines of code being commented/uncommented then (with no defined authority on the matter) vote that you should change to Blinn-Phong and merge this when it passes. If we have strong feelings from @asnt that Phong is better for VisPy then we can switch it later. Thoughts?

My kid is sick today so I won't be working much. If you're ok making the change and the tests pass then feel free to merge.

**Comment by almarklein** [comment]
There are small differences, but it's hard to tell which one is "better" from these images. [This image](https://learnopengl.com/img/advanced-lighting/advanced_lighting_comparrison.png) from the post I referenced above shows the differences more specifically.

Phong:
image
Blinn-Phong:
image

**Comment by almarklein** [comment]
The code is just 2 lines each. The difference in performance is probably negligible on today's hardware. I'll wait for @asnt and otherwise merge this evening.

**Comment by djhoese** [comment]
> but it's hard to tell which one is "better" from these images.

Yeah, swapping between the two images quickly and I think I could convince myself that either of them is better or more "realistic" for some reason.

**Pull request merged**
